### ISSUE TITLE
Why is discussion of Mocha.jl not active than expectation?

### ISSUE THREAD

**Issue by jskDr** [pluskid/Mocha.jl#157]
I am not sure whether this is right place to elaborate this topic or it should be discussed in the place of main Julia discussion. Assuming here is the right one. I would say that whenever I come to visit here I am so wondering why it is not so active in this place than my expectation. 

Yes, this place is relatively so active while most other computation discussions become just dead after starting even in a few days. However, let us consider that this is a period of machine learning and especially deep learning and neural networks which are now so emerging in everywhere. I am not sure whether it is correct thought or not but I think Julia should be judged by showing that it is a good language to support machine learning. If not well supportable, people might not think Julia is a practical language highly probably. 

Julia is obviously a very good and powerful language. However if it is not so much practical there is some probability it eventually becomes similar to EMACS and LISP which are the best tools but majorly for experts and their followers. 

So, I hope that if Julia is not sufficient or over implemented to support machine learning such as Mocha.jl it is also good idea even to prune the currently implemented functions as well as to add new functions supporting machine learning. By doing these activities, I would expect that in Julia machine learning  discussions like in this place should be highly active. 

I am the one who hate CLASS but enjoy TYPE. Hence, I hope Julia to be a general and powerful language in computation and machine learning fields.

**Comment by pluskid** [comment]
There are several reasons. There are also recently discussions on the [Roadmap Issue of JuliaStats](https://github.com/JuliaStats/Roadmap.jl/issues/11#issuecomment-147104338). You are free to join the conversation if you have something to say.

* First of all, I think Julia is still at its early stage, though it has been several years of development. If you compare the number of packages in Julia vs in PyPI, the latter is 100 times more than the former. So the whole ecosystem is not built within 1 month or a year. As a results, there are just not enough users to participate, either in discussion or in maintaining packages.
* The fact that Julia being at its early stage also means some functionalities are still not as good, even though the language itself is already quite nice to work with. For example, we still need to use [devectorize](https://github.com/lindahua/Devectorize.jl) explicitly to avoid allocating unnecessary temporary large arrays. Another example is related to deep learning. There are cuda array support in Matlab, Python, Lua, etc. But we do not have a full featured Array type for GPU yet.
* Thirdly, while machine learning researchers find Julia and think it is a perfect match, the primary target of the language itself is not machine learning or statistics. The core team has broader focus on other fields of numerical / scientific computing, engineering, etc.

**Comment by malmaud** [comment]
I'm also not sure that activity on Github is a good proxy for success. [Torch's github](https://github.com/torch/torch7/issues) only has about 2x the activity of Mocha's, judging by number of issues, and Torch is being actively developed and promoted by full-time engineers at Google and Facebook. Considering that Mocha is primarily developed by a few volunteers as a side job, that seems pretty good. 

Julia is also on the radar of major machine learning groups, including Zoubin Gharamani's, Ryan Adams', and Deniz Yuret's. There are people actively working on Julia GPU programming - most of NVidia's CUDA libraries are wrapped or in the process of being wrapped now, and there are ambitious efforts to get Julia functions to directly compile to CUDA kernels underway.

**Comment by jskDr** [comment]
@pluskid and @malmaud, if Julia is early stage and also machine learning in Julia is under developing to be as good as Caffe and Torch7, I suggest for Mocha.jl/Julia to be differentiate from them instead of following them. 

If Torch7 and Caffe are persuading mainly the implementation of deep-learning using already well established computer languages such as C and C++,  why Mocha.jl focus on initiating post deep-learning whatever it is. Although there are some discussions and visions [1,2] on post deep-learning It is still not clear. However, around time Julia being 1.0 the vision of it will be much clearer than now. Hence, I am not sure but this differentiation strategy seems to be eventually one of the winning strategies over all other approaches. 
[1] Cognitive Computing: Beyond Deep Learning, http://www.enterrasolutions.com/2015/03/cogntive-computing-beyond-deep-learning.html, 2015
[2] Cognitive computing, http://www.research.ibm.com/cognitive-computing

Regarding post deep-learning including advanced and beyond deep learning, my thought is as follows. 

First, the combination of Lasso-like learning on the top of deep learning can be one direction of it and this direction is already realizing by adding direct Lasso types or modified Lasso types of regularization methods such as L1, L2, dropout methods. Although there have been some successful results of the combinatoric trials [3], it should be still further investigated to much complete it since it is belong to almost NP-hard problems for nonlinear classification or regression cases. 
[3] N. Srivastava, G. Hinton, A. Krizhevsky, I. Sutskever, and R. Salakhutdinov, “Dropout: A Simple Way to Prevent Neural Networks from Overfitting,” J. Mach. Learn. Res., vol. 15, no. 1, pp. 1929–1958, Jan. 2014.

Second, cognitive learning is considered as the next generation of machine learning while Lasso-like learning and deep-learning is a state of art technology. However, ideal cognitive learning by the combination of deep learning and Bayesian reasoning seems to be long way to go as far as I know. Before it, more practical and concrete vision of the new learning is needed to be establish so as to be followed by associated researchers and engineers. Let us say it is 'smart learning' to distinguish with 'ideal' cognitive learning. 

In short, if we discuss here not only how to implement deep-learning using baby Julia but also what is the next generation of machine learning and how to prepare it with futuristic Julia, I think Mocha.jl/Julia can be greatly differentiated from currently 'de facto' machine learning platforms of Torch7 and Mocha. What do you, fans of Mocha.jl, think of this approach?

p.s. Recently I guess I got an small hint of 'smart learning' during trial of my material informatics study. In the study, I had to test a bunch of new molecules, say need to predict properties of more than hundred thousand molecules while I only have a few dozens of experimental data which are, unfortunately, barely correlated with the test data in average. These problems can not be solved by deep learning because of insufficient number of the training data and also hard to be solved by Lasso-like learning because of low correlation between the training and the test data. I might now found one preliminary solution, which shows doubtfully high performance, i.e., r square as 0.90 for a certain property. I am really not sure whether this experience is a glimpse of 'smart learning' but it is sure to be different from current activities of machine learning. So, I start to implement my trial using powerful and flexible language of Julia based on the machine learning engine of Mocha.jl to be used for more general cases . Hope your helps a lot. 

Sungjin (James) Kim
sungjinkim@fas.harvard.edu

**Comment by kleinash** [comment]
Putting down my 2c.. 

I like Mocha. I would like to carry on using it for what I am trying to do deep learning wise. I like that its built directly on cudNN, making it easy to use with GPU backend. That I can specify my own layers and that it runs! Its simple, not layered with so much that its hard to see whats happening (like in theano). I like Julia, and that the source code is in Julia. 

I would really like this area to be more active - because I am currently trying to iron out issues that I think have to do with the constant updating of Julia - I get that this is in its infancy and thats to be expected, but I like being able to work with the developers to get things running. 

I am unclear on how to allocate memory when using the GPU back end - especially when my images start getting larger. But I am fully in support of Mocha and Julia.

**Comment by malmaud** [comment]
Mocha is essentially deprecated in favor of mxnet.jl now.

**Comment by kleinash** [comment]
@mxnet 
thanks for that
